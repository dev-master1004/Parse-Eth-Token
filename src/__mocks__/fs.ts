import { basename } from 'path';

export const promises = {
  async readdir() {
    return ['bar.json', 'baz.json'];
  },

  async readFile(path: string) {
    const file = basename(path);

    // Few different test cases
    let data;
    switch (file) {
      case 'foo.json':
        data = {
          address: '0x0',
          decimals: 18,
          symbol: 'FOO',
          name: 'Foo'
        };
        break;
      case 'bar.json':
        data = {
          address: '0x0000000000000000000000000000000000000000',
          decimals: 18,
          symbol: 'BAR',
          name: 'Bar'
        };
        break;
      case 'baz.json':
        data = {
          address: '0x0000000000000000000000000000000000000001',
          decimals: 18,
          symbol: 'BAZ',
          name: 'Baz'
        };
        break;
      case 'invalid.json':
        return 'I am invalid JSON';
    }

    return JSON.stringify(data);
  },

  writeFile: jest.fn().mockImplementation(() => Promise.resolve()),

  mkdir: jest.fn().mockImplementation(() => Promise.resolve()),

  access: jest.fn().mockImplementation(async () => {
    const error = new Error() as NodeJS.ErrnoException;
    error.code = 'ENOENT';

    throw error;
  })
};
