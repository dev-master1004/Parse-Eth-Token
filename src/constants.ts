import { resolve } from 'path';
import { tmpdir } from 'os';
import { number, object, string } from '@hapi/joi';

type ValueOf<T> = T[keyof T];

export const VERSION = '1.0.0';
export const REPO_URL = 'https://github.com/ethereum-lists/tokens.git';
export const OUTPUT_PATH = resolve(tmpdir(), 'ethereum-lists/tokens');

/**
 * `name` taken from `ethereum-lists/tokens` (https://github.com/ethereum-lists/tokens/tree/master/tokens)
 * `chainId` taken from https://chainid.network/
 */
export const NETWORKS = [
  {
    name: 'ella',
    chainId: 64
  },
  {
    name: 'esn',
    chainId: 31102
  },
  {
    name: 'etc',
    chainId: 61
  },
  {
    name: 'eth',
    chainId: 1
  },
  {
    name: 'gor',
    chainId: 5
  },
  {
    name: 'kov',
    chainId: 42
  },
  {
    name: 'rin',
    chainId: 4
  },
  {
    name: 'rop',
    chainId: 3
  },
  {
    name: 'ubq',
    chainId: 8
  }
];

export const NETWORK_NAMES = NETWORKS.map(network => network.name);

export const SOCIAL_SCHEMA = object({
  blog: string()
    .empty('')
    .default(''),
  chat: string()
    .empty('')
    .default(''),
  facebook: string()
    .empty('')
    .default(''),
  forum: string()
    .empty('')
    .default(''),
  github: string()
    .empty('')
    .default(''),
  gitter: string()
    .empty('')
    .default(''),
  instagram: string()
    .empty('')
    .default(''),
  linkedin: string()
    .empty('')
    .default(''),
  reddit: string()
    .empty('')
    .default(''),
  slack: string()
    .empty('')
    .default(''),
  telegram: string()
    .empty('')
    .default(''),
  twitter: string()
    .empty('')
    .default(''),
  youtube: string()
    .empty('')
    .default('')
});

const DEPRECATED_TOKEN_SCHEMA = object({
  new_address: string()
});

/**
 * Validation schema.
 */
export const TOKEN_SCHEMA = object({
  address: string()
    .regex(/^0x[a-fA-F0-9]{40}$/)
    .required(),
  decimals: number()
    .min(0)
    .required(),
  deprecation: DEPRECATED_TOKEN_SCHEMA,
  name: string().required(),
  social: SOCIAL_SCHEMA,
  symbol: string().required(),
  website: string()
    .empty('')
    .default('')
}).options({ stripUnknown: true });

/**
 * Raw token data that is loaded from the JSON files.
 */
export interface TokenSocialMetadata {
  blog?: string;
  chat?: string;
  facebook?: string;
  forum?: string;
  github?: string;
  gitter?: string;
  instagram?: string;
  linkedin?: string;
  reddit?: string;
  slack?: string;
  telegram?: string;
  twitter?: string;
  youtube?: string;
}
/**
 * Raw token data that is loaded from the JSON files.
 */
export interface RawToken {
  address?: string;
  decimals?: number | string;
  deprecation?: any;
  name?: string;
  social?: TokenSocialMetadata;
  symbol?: string;
  website?: any;
}

/**
 * Parsed token data.
 */
export interface Token {
  address: string;
  decimals: number;
  deprecation?: any;
  name: string;
  newSymbol?: any;
  social?: TokenSocialMetadata;
  symbol: string;
  uuid: string;
  website?: any;
}
