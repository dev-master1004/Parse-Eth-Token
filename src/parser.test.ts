import { promises as fs } from 'fs';
import { resolve } from 'path';
import mkdirp from 'mkdirp';
import {
  addUniqueId,
  checkNetworks,
  createOutputFolder,
  fixDuplicates,
  parseJsonFile,
  parseTokenFiles,
  sortTokens,
  validateTokenData,
  writeToDisk
} from './parser';
import { NETWORK_NAMES, RawToken, Token } from './constants';

jest.mock('fs');

afterEach(() => {
  jest.clearAllMocks();
});

describe('checkNetworks', () => {
  it('checks for valid networks', () => {
    expect(() => checkNetworks(NETWORK_NAMES)).not.toThrow();
  });

  it('throws an error on invalid networks', () => {
    expect(() => checkNetworks(['foo', 'bar'])).toThrow();
  });
});

describe('parseJsonFile', () => {
  it('reads and parses a JSON file', async () => {
    const file = await parseJsonFile<RawToken>('foo.json');

    expect(file).toStrictEqual({
      address: '0x0',
      decimals: 18,
      symbol: 'FOO',
      name: 'Foo'
    });
  });

  it('throws on invalid JSON', async () => {
    await expect(parseJsonFile('invalid.json')).rejects.toThrow();
  });
});

describe('validateTokenData', () => {
  it('strips unknown fields', () => {
    const token = {
      address: '0x0000000000000000000000000000000000000000',
      decimals: 18,
      symbol: 'FOO',
      name: 'Foo',
      foo: 'bar'
    };

    expect(validateTokenData(token)).not.toHaveProperty('foo', 'bar');
  });

  it('throws an error on invalid input', () => {
    expect(() => validateTokenData({})).toThrow();
    expect(() =>
      validateTokenData({
        address: '0x0',
        decimals: 18,
        symbol: 'FOO',
        name: 'Foo'
      })
    ).toThrow();
    expect(() =>
      validateTokenData({
        address: '0x0000000000000000000000000000000000000000',
        decimals: -1,
        symbol: 'FOO',
        name: 'Foo'
      })
    ).toThrow();
  });
});

describe('parseTokenFiles', () => {
  it('parses multiple token files', async () => {
    const tokens = await parseTokenFiles('/', []);

    expect(tokens.length).toBe(2);
    expect(tokens[0]).toStrictEqual({
      address: '0x0000000000000000000000000000000000000000',
      decimals: 18,
      name: 'Bar',
      symbol: 'BAR',
      website: ''
    });
    expect(tokens[1]).toStrictEqual({
      address: '0x0000000000000000000000000000000000000001',
      decimals: 18,
      name: 'Baz',
      symbol: 'BAZ',
      website: ''
    });
  });

  it('excludes addresses from the list', async () => {
    const tokens = await parseTokenFiles('/', ['0x0000000000000000000000000000000000000000']);

    expect(tokens.length).toBe(1);
    expect(tokens[0]).toStrictEqual({
      address: '0x0000000000000000000000000000000000000001',
      decimals: 18,
      name: 'Baz',
      symbol: 'BAZ',
      website: ''
    });
  });
});

describe('addUniqueId', () => {
  it('adds a deterministic unique ID to tokens', () => {
    const tokens = [
      {
        address: '0x0000000000000000000000000000000000000000',
        symbol: 'FOO',
        decimals: 18,
        name: 'Foo'
      },
      {
        address: '0x0000000000000000000000000000000000000001',
        symbol: 'BAR',
        decimals: 18,
        name: 'Bar'
      },
      {
        address: '0x0000000000000000000000000000000000000002',
        symbol: 'BAZ',
        decimals: 18,
        name: 'Baz'
      },
      {
        address: '0x0000000000000000000000000000000000000003',
        symbol: 'QUX',
        decimals: 18,
        name: 'Qux'
      }
    ];

    expect(addUniqueId(tokens as Token[], 1)).toMatchSnapshot();
    expect(addUniqueId(tokens as Token[], 2)).toMatchSnapshot();
  });
});

describe('fixDuplicates', () => {
  it('changes the symbol for duplicate tokens', () => {
    const tokens = [
      {
        address: '0x0000000000000000000000000000000000000000',
        symbol: 'FOO',
        decimals: 18,
        name: 'Foo',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000001',
        symbol: 'FOO',
        decimals: 18,
        name: 'Foo',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000002',
        symbol: 'FOO',
        decimals: 18,
        name: 'Foo',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000003',
        symbol: 'BAR',
        decimals: 18,
        name: 'Bar',
        uuid: ''
      }
    ];

    const sorted = fixDuplicates(tokens);

    expect(sorted.length).toBe(4);
    expect(sorted[0]).toHaveProperty('symbol', 'FOO');
    expect(sorted[1]).toHaveProperty('symbol', 'FOO (1)');
    expect(sorted[2]).toHaveProperty('symbol', 'FOO (2)');
    expect(sorted[3]).toHaveProperty('symbol', 'BAR');
  });
});

describe('sortTokens', () => {
  it('sorts tokens alphabetically based on the symbol', () => {
    const tokens = [
      {
        address: '0x0000000000000000000000000000000000000000',
        decimals: 18,
        symbol: 'FOO',
        name: 'Foo',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000000',
        decimals: 18,
        symbol: 'BAR',
        name: 'Bar',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000000',
        decimals: 18,
        symbol: 'BAZ',
        name: 'Baz',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000000',
        decimals: 18,
        symbol: 'QUX',
        name: 'Qux',
        uuid: ''
      }
    ];

    const sorted = sortTokens(tokens);

    expect(sorted).toHaveLength(4);
    expect(sorted[0]).toHaveProperty('symbol', 'BAR');
    expect(sorted[1]).toHaveProperty('symbol', 'BAZ');
    expect(sorted[2]).toHaveProperty('symbol', 'FOO');
    expect(sorted[3]).toHaveProperty('symbol', 'QUX');
  });
});

describe('createOutputFolder', () => {
  it('creates the output foldder', async () => {
    await expect(createOutputFolder('foo')).resolves.toBeUndefined();

    const mkdir = mkdirp.sync as jest.MockedFunction<typeof mkdirp.sync>;

    expect(mkdir.mock.calls[0][0]).toBe('foo');
  });

  it('throws an error if the error code is not ENOENT', async () => {
    const access = fs.access as jest.MockedFunction<typeof fs.access>;
    access.mockImplementationOnce(async () => {
      const error = new Error() as NodeJS.ErrnoException;
      error.code = 'EFOO';

      throw error;
    });

    await expect(createOutputFolder('foo')).rejects.toThrow();
  });
});

describe('writeToDisk', () => {
  it('writes the tokens to a file', async () => {
    const tokens = [
      {
        address: '0x0000000000000000000000000000000000000000',
        symbol: 'FOO',
        decimals: 18,
        name: 'Foo',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000001',
        symbol: 'FOO',
        decimals: 18,
        name: 'Foo',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000002',
        symbol: 'FOO',
        decimals: 18,
        name: 'Foo',
        uuid: ''
      },
      {
        address: '0x0000000000000000000000000000000000000003',
        symbol: 'BAR',
        decimals: 18,
        name: 'Bar',
        uuid: ''
      }
    ];

    const json = JSON.stringify(tokens, null, 2);

    const writeFile = fs.writeFile as jest.MockedFunction<typeof fs.writeFile>;

    await writeToDisk(tokens, '/', 'foo.json');
    expect(writeFile).toHaveBeenCalledTimes(1);
    expect(writeFile.mock.calls[0][0]).toBe(resolve('/', 'foo.json'));
    expect(writeFile.mock.calls[0][1]).toBe(json);
    expect(writeFile.mock.calls[0][2]).toBe('utf8');
  });
});
