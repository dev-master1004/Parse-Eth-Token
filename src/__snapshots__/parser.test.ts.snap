// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addUniqueId adds a deterministic unique ID to tokens 1`] = `
Array [
  Object {
    "address": "0x0000000000000000000000000000000000000000",
    "decimals": 18,
    "name": "Foo",
    "symbol": "FOO",
    "uuid": "163503a1-bc37-52d8-8325-0a2b5869563d",
  },
  Object {
    "address": "0x0000000000000000000000000000000000000001",
    "decimals": 18,
    "name": "Bar",
    "symbol": "BAR",
    "uuid": "f653068e-ed65-5359-a950-b0d5cb3c7977",
  },
  Object {
    "address": "0x0000000000000000000000000000000000000002",
    "decimals": 18,
    "name": "Baz",
    "symbol": "BAZ",
    "uuid": "825e38a3-3ad3-5cd4-ae41-4b36bb82ca52",
  },
  Object {
    "address": "0x0000000000000000000000000000000000000003",
    "decimals": 18,
    "name": "Qux",
    "symbol": "QUX",
    "uuid": "b2f15f5a-c2db-5b87-88d3-a7c92d54b52d",
  },
]
`;

exports[`addUniqueId adds a deterministic unique ID to tokens 2`] = `
Array [
  Object {
    "address": "0x0000000000000000000000000000000000000000",
    "decimals": 18,
    "name": "Foo",
    "symbol": "FOO",
    "uuid": "b79e2448-8d6a-5a4c-8b19-b06847d10461",
  },
  Object {
    "address": "0x0000000000000000000000000000000000000001",
    "decimals": 18,
    "name": "Bar",
    "symbol": "BAR",
    "uuid": "9227ccb6-bcd8-532e-812c-389ecc8804f5",
  },
  Object {
    "address": "0x0000000000000000000000000000000000000002",
    "decimals": 18,
    "name": "Baz",
    "symbol": "BAZ",
    "uuid": "7f875eb6-2bbe-51e1-a303-9f81a9b38852",
  },
  Object {
    "address": "0x0000000000000000000000000000000000000003",
    "decimals": 18,
    "name": "Qux",
    "symbol": "QUX",
    "uuid": "7e57e5b3-4967-501f-9ddc-9e906b11b49d",
  },
]
`;
